pkg/cmd/admin/image/verify-signature.go

```diff
@@ -0,0 +1,230 @@
+		mechanism, err = newGPGSigningMechanismInDirectory("")
+	} else {
+		mechanism, _, err = newEphemeralGPGSigningMechanism(o.PublicKey)
+	}
+	if err != nil {
+		return "", nil, err
+	}
+	defer mechanism.Close()
+	content, identity, err := mechanism.Verify(signature)
+	if err != nil {
+		return "", nil, err
+	}
+	return string(identity), content, nil
+}
+
+// verifySignatureContent verifies that the signature content matches the given image.
+// TODO: This should be done by calling the 'containers/image' library in future.
+func (o *VerifyImageSignatureOptions) verifySignatureContent(content []byte) error {
+	// TODO: The types here are just to decompose the JSON. The fields should not change but
+	// we need to use containers/image library here to guarantee compatibility in future.
+	type criticalImage struct {
+		Digest string `json:"docker-manifest-digest"`
+	}
+	type critical struct {
+		Image criticalImage `json:"image"`
+	}
+	type message struct {
+		Critical critical `json:"critical"`
+	}
+	m := message{}
+	if err := json.Unmarshal(content, &m); err != nil {
+		return err
+	}
+	if o.InputImage != m.Critical.Image.Digest {
+		return fmt.Errorf("signature is valid for digest %q not for %q", m.Critical.Image.Digest, o.InputImage)
+	}
+	return nil
+}
+
+// removeImageSignature removes the current image signature from the Image object by
+// erasing all signature fields that were previously set (when image signature was
+// previously verified).
+func (o *VerifyImageSignatureOptions) removeImageSignature(s *imageapi.ImageSignature) {
+	s.Conditions = []imageapi.SignatureCondition{}
+	s.IssuedBy = nil
+}
+
+func (o *VerifyImageSignatureOptions) Run() error {
+	img, err := o.Client.Images().Get(o.InputImage)
+	if err != nil {
+		return err
+	}
+	if len(img.Signatures) == 0 {
+		return fmt.Errorf("%s does not have any signature", img.Name)
+	}
+
+	for i, s := range img.Signatures {
+		// If --remove is specified, just erase the existing signature verification for all
+		// signatures.
+		// TODO: This should probably need to handle removal of a single signature.
+		if o.Remove {
+			o.removeImageSignature(&img.Signatures[i])
+			continue
+		}
+
+		// Verify the signature against the public key
+		signedBy, content, signatureErr := o.verifySignature(s.Content)
+		if signatureErr != nil {
+			fmt.Fprintf(o.ErrOut, "%s signature cannot be verified: %v\n", o.InputImage, signatureErr)
+		}
+
+		// Verify the signed message content matches with the provided image id
+		var signatureContentErr error
+		if signatureContentErr = o.verifySignatureContent(content); signatureContentErr != nil {
+			fmt.Fprintf(o.ErrOut, "%s signature content cannot be verified: %v\n", o.InputImage, signatureContentErr)
+		}
+
+		// If we have error and --confirm then remove the signature from the image which
+		// changes the image to "unverified" state.
+		if signatureErr != nil || signatureContentErr != nil {
+			o.removeImageSignature(&img.Signatures[i])
```

Comment:
Can we record the error in the signature as as a condition `{"Type": "Trusted", "Status": "false", "Reason": "error", "Message": err}`?